### PR TITLE
libnet/ipam: Lazily sub-divide pools into subnets

### LINKED ISSUE #40275: IPv6 address pool subnet smaller than /80 causes dockerd to consume all available RAM
**Description**

It is documented that IPv6 pool "should" be at least /80 so that MAC address can fit in the last 48 bits.

Using a default-address-pools size larger than 80 causes dockerd to consume too much RAM - the longer the prefix (smaller subnet), the more RAM dockerd will use:
- At /81 - /90 range the RAM usage increase is negligible in the range of few GB.
- At /94 - /96 range the RAM usage is in the tens to hundreds of GB.

The pool prefix length could be set to larger by 80 by a typing error or a mistake and if this leads to dockerd consuming copious amounts of RAM will cause the administrator to possibly lose time troubleshooting the situation.

It looks like prefix length like /96 is totally unusable and dockerd should refuse to start instead of starting to allocate ridiculous amounts of RAM.

At minimum a warning message should be printed.

**Steps to reproduce the issue:**
1. Set IPv6 address pool prefix length longer than 80:
```
  "default-address-pools": [
    { "base": "192.0.2.0/16", "size": 24 },
    { "base": "2001:db8:1:1f00::/64", "size": 96 }
  ],
```
2. Start Docker.
3. Watch the server grind to a halt and kernel OOM killer being invoked.

**Describe the results you received:**
dockerd consumes very large amounts of RAM (tens of GB).

**Describe the results you expected:**
Either IPv6 pool prefix lengths longer than 80 should work, or dockerd should refuse start with a configuration that cannot be used.

At minimum a warning message should be printed for prefix lengths longer than 80.

The documentation does not mention the RAM usage effect either:
```
The subnet for Docker containers should at least have a size of /80, so that an IPv6 address can end with the container’s MAC address and you prevent NDP neighbor cache invalidation issues in the Docker layer.
```

**Additional information you deem important (e.g. issue happens only occasionally):**
100% reproducible.

**Output of `docker version`:**

```
Docker version 19.03.5, build 633a0ea
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 19.03.5
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: fluentd
 Cgroup Driver: systemd
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
  selinux
 Kernel Version: 3.10.0-1062.4.3.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 24
 Total Memory: 47.15GiB
 Name: docker.domain
 ID: xxx
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### LINKED ISSUE #42801: splitNetwork and addIntToIP fails for IPv6 addresses on /64+ size pools
**Description**

When attempting to create the pools for IPv6 addresses ranges, the output of `splitNetwork` produces the same CIDR N times rather than the N different CIDRs if the output network size is >= 64 bits. This is due to the value passed to `addIntToIP` being a 64 bit int, which is insufficient to contain the required value to offset IPv6 ranges.

**Steps to reproduce the issue:**

1. Add an IPv6 range to `default-address-pools` in config and enable IPv6.

For example:
```
   "default-address-pools": [
      < IPv4 pools >
      {
        "base": "fd00:0000:0000::/48",
        "size": 64
      }
    ]
```

2. Attempt to create two IPv6 networks, the second will fail with `could not find an available, non-overlapping IPv6 address pool among the defaults to assign to the network`, suggesting there is only one pool available

**Describe the results you received:**

Only a single IPv6 pool is available per entry in `default-address-pools`

**Describe the results you expected:**

The output of splitting `fd00:0000:0000::/48` into `/64` networks should yield `2^16` pools, not one.

**Additional information you deem important (e.g. issue happens only occasionally):**

This is caused by [`splitNetwork` and `addIntToIP`](https://github.com/moby/moby/blob/8684f482e4cb84fee269d8e91176ca8e1afcc218/libnetwork/ipamutils/utils.go#L92-L135), which works OK for IPv4 but fails for IPv6 networks of `/64` or larger due to the `ordinal` argument to  `addIntToIP` being a 64 bit int, which ends up being zero when using a `/64` or larger. Crucially, the statement `uint(i<<s)` yields zero here, since `s > 64`

For example, `fmt.Println(splitNetworks([]*NetworkToSplit{{"fd00:0000:0000::/48", 50}}))` yields the same CIDR repeated 4 times rather than 4 separate CIDRs:

```
[fd00::/50 fd00::/50 fd00::/50 fd00::/50]
```

To fix this, those two functions need to be amended to allow values up to 128 bits to be added to an IP, perhaps with an `ordinalUpper` and `ordinalLower` as two 64 bit uints.

I believe this issue may be the underlying cause of https://github.com/moby/moby/issues/41438

**Output of `docker version`:**

```
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.11.6
 Git commit:        4c52b90
 Built:             Sun, 21 Feb 2021 18:18:35 +0100
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.6
  Git commit:       4c52b90
  Built:            Sun Feb 21 17:18:35 2021
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 27
 Running: 27
 Paused: 0
 Stopped: 0
Images: 31
Server Version: 18.09.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 1.0.0~rc6+dfsg1-3
init version: v0.18.0 (expected: fec3683b971d9c3ef73f284f176672c44b448662)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.19.0-14-amd64
Operating System: Debian GNU/Linux 10 (buster)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.4GiB
Name: 
ID: GXW4:VDGE:BS4J:AY3S:5DHN:COS3:TCUC:JOBN:KRPG:V6MW:YO37:A2GO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 191
 Goroutines: 192
 System Time: 2021-08-30T22:44:47.757633368Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
None

### PR DESCRIPTION
**Pull request by akerouanton**
I bundled these two commits together as the second one is easier to test with changes done in the first one. I can still break this PR into two parts and rewrite the unit test of the second commit to test it with the current master branch. Let me know if you prefer that (eg. if you want to cherry-pick this change into older releases or because it's a big diff to review).

**- 1st commit**

This commit resolves #40275 by implementing a custom iterator
named NetworkSplitter. It splits a set of NetworkToSplit into smaller
subnets on-demand by calling its Get method.

Prior to this change, the list of NetworkToSplit was split into smaller
subnets when ConfigLocalScopeDefaultNetworks or
ConfigGlobalScopeDefaultNetworks were called or when ipamutils package
was loaded. When one of the Config function was called with an IPv6 net
to split into small subnets all the available memory was consumed. For
instance, fd00::/8 split into /96 would take ~5*10^27 bytes.

Although this change trades memory consumption for computation cost, the
NetworkSplitter is used by libnetwork/ipam package in such a way that it
only have to compute the address of the next subnet. When
NetworkSplitter reach the end of NetworkToSplit, it's resetted by
libnetwork/ipam only if there were some subnets released beforehand. In
such case, ipam package might iterate over all the subnets before
finding one available subnet. This is the worst-case but it shall not be
really impactful as either many subnets exists (more than one host can
handle) or not so many subnets exists and full iteration over
NetworkSplitter is fast.

**- 2nd commit**

Prior to this change, IPv6 subnets generation was broken because of a
bitwise shift overflowing uint64.

To solve this issue, addIntToIP has been changed to take an addend and
computes two ordinals applied to respectively the lower 64 bits and the
upper 64 bits of IPv6 addresses.

Nothing change for IPv4 (ie. upperOrdinal is always 0).

**- How to verify it**

I added unit tests to confirm these changes are okay. Moreover, I tried to run dockerd with following config:

```json
{
	"log-level": "info",
	"ipv6": true,
	"experimental": true,
	"ip6tables": true,
	"fixed-cidr-v6": "2a01:e34:xxxx:yyyy::/64",
	"default-address-pools": [
		{ "base": "172.17.0.0/16", "size": 16 },
		{ "base": "172.18.0.0/16", "size": 16 },
		{ "base": "2a01:e34:xxxx:yyyy::/64", "size": 120 }
	]
}
```

When starting dockerd v20.10.10 with this config, the daemon crashes with this panic message message: `runtime error: makeslice: cap out of range`. When starting a patched version of dockerd, it doesn't crash.

**- Description for the changelog**

* Generate split subnets on-demand to not consume all the available memory when small IPv6 subnets are created from big subnets (fixes #40275)
* Properly shift bytes of IPv6 subnets (fixes #42801)

### CODE REVIEW

**Comment by Xyaren** [comment]
Hey, awesome that somebody is giving ipv6 some love 👍 
Any chance this is fixing #41438 aswell ?

**Comment by akerouanton** [comment]
@Xyaren I still need a bit of time to test it properly but yeah, it should fix #41438 as (I believe) this is a consequence of the uint64 overflow :wink: Thanks for mentioning this issue :slightly_smiling_face:

**Comment by akerouanton** [comment]
@thaJeztah PTAL :slightly_smiling_face: (I don't know who else I should ping. Maybe arkodg?)

**Comment by Xyaren** [comment]
So.. about 5 month later...
Is there anyone able to look at this please ?

**Comment by thaJeztah** [comment]
@akerouanton looks like some failures in a test (may need some extra options in it);

```
=== Failed
=== FAIL: github.com/docker/docker/daemon/config TestDaemonConfigurationMergeDefaultAddressPools/empty_config_file (0.00s)
    config_test.go:155: assertion failed: cannot handle unexported field at {[]*ipamutils.NetworkToSplit}[0].ipVersion:
        	"github.com/docker/docker/libnetwork/ipamutils".NetworkToSplit
        consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported
    --- FAIL: TestDaemonConfigurationMergeDefaultAddressPools/empty_config_file (0.00s)
=== FAIL: github.com/docker/docker/daemon/config TestDaemonConfigurationMergeDefaultAddressPools/config_file (0.00s)
    config_test.go:165: assertion failed: cannot handle unexported field at {[]*ipamutils.NetworkToSplit}[0].ipVersion:
        	"github.com/docker/docker/libnetwork/ipamutils".NetworkToSplit
        consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported
    --- FAIL: TestDaemonConfigurationMergeDefaultAddressPools/config_file (0.00s)
````

**Comment by akerouanton** [comment]
@thaJeztah Ah, looks like a bad rebase. Should be fixed now.

**Comment by max-wittig** [comment]
@thaJeztah Any chance to have another look at this. Would be really nice to have that fixed!

**Comment by max-wittig** [comment]
@thaJeztah @cpuguy83 Friendly ping 😄 . We've added IPv6 networking support to the gitlab-runner, but we have no way to use it as IPv6 networking in Docker is still broken, when used in this way.

Is there any way I can support the efforts here?

File: libnetwork/ipamutils/utils.go
Comment (cpuguy83) [inline libnetwork/ipamutils/utils.go:187]: I think we could have `Get()` return `nil` when it is used up instead of having this extra function.
`Next` itself is a bit of a misnomer since it's not advancing anything just checking if `Get()` would return something.
It seems like all the call sites could do `for net := splitter.Get(); net != nil { ... }` instead.

File: libnetwork/ipamutils/utils.go
Comment (cpuguy83) [inline libnetwork/ipamutils/utils.go:158]: Maybe a note that this is primarily here for testing purposes?

File: libnetwork/ipamutils/utils.go
Comment (cpuguy83) [inline libnetwork/ipamutils/utils.go:111]: I would feel better if we had had a mutex in here or at least an explicit call-out that additional synchronization may be required and how it is used safely.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: libnetwork/ipamutils/utils.go
Comment (akerouanton) [reply libnetwork/ipamutils/utils.go:187]: Although in most cases checking the return value of `Get` would do the job, I think it'd make the code resetting the iterator harder to understand (cf. [see here](https://github.com/moby/moby/pull/43033/files#diff-9265e3cef0c7e0f9b3373f21ede0e1a56f625a8d86785655e1229636f9337979R432-R435)). However, I think you're right: `Next` is badly named. Maybe I can rename it into `HasNext` or something like that. What do you think?

Comment (akerouanton) [reply libnetwork/ipamutils/utils.go:187]: It just feels better to let `Get` return nil if there's no more subnet to return. As such, I renamed `Next` into `HasNext` to make it more convenient to check if the splitter needs to be reset and removed all calls to `Next` except for that particular case. Let me know if you have other suggestions here.

**Comment by thaJeztah** [comment]
@akerouanton Looks like the linter caught a typo;

```
libnetwork/ipamutils/utils.go:160:19: `primarly` is a misspelling of `primarily` (misspell)
// This method is primarly used to duplicate NetworkSplitter in tests.
                  ^
```

And there's one test-failure (haven't looked if its a legit failure, or a test that needs updating);

```
=== FAIL: github.com/docker/docker/libnetwork/ipamutils TestInitPoolWithIPv6 (0.00s)
    utils_test.go:154: assertion failed: <nil> (string) != fd00:0:ffff:ffff:ffff:ffff::/96 (string)
    utils_test.go:158: assertion failed: <nil> (string) != fd00:0:ffff:ffff:ffff:ffff::/96 (string)
```

**Comment by akerouanton** [comment]
@thaJeztah The failing test looks legit but I didn't had time to debug it. I'll fix that later and ping you when it's done 😉

**Comment by akerouanton** [comment]
@thaJeztah I'm not sure why the integration-cli job is failing. I don't reproduce this error on my computer. Could you take a look please?

**Comment by thaJeztah** [comment]
Hm... not sure; would have to check why it fails. I can give it a kick (in case there's some flakiness in that test, or (e.g.) some other daemon was still running).

Copied the error below (in case we need to revisit)

```
=== Failed
=== FAIL: amd64.integration-cli TestDockerNetworkSuite/TestDockerNetworkIPAMMultipleNetworks (15.30s)
    docker_cli_network_unix_test.go:588: assertion failed: 
        Command:  /usr/local/cli/docker network create test3
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   Error response from daemon: could not find an available, non-overlapping IPv4 address pool among the defaults to assign to the network
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    docker_cli_network_unix_test.go:46: [d17053f569bde] daemon is not started
    --- FAIL: TestDockerNetworkSuite/TestDockerNetworkIPAMMultipleNetworks (15.30s)

=== FAIL: amd64.integration-cli TestDockerNetworkSuite (769.49s)
```

**Comment by neersighted** [comment]
It looks like a legitimate test failure. In addition, the commits are unsigned and you have verification turned on. Could you make sure your commits are signed after you solve the test issue?

**Comment by neersighted** [comment]
@akerouanton some test failures are known/from a stale head. Can you please rebase onto master?

**Comment by akerouanton** [comment]
@neersighted Already did that before my last force-push :wink:

**Comment by neersighted** [comment]
Hmm, odd we're still seeing the buildkit-related ones then:

```
    --- FAIL: TestIntegration/TestMountRWCache/worker=dockerd/frontend=client (1.87s)
    --- FAIL: TestIntegration/TestMountRWCache/worker=dockerd/frontend=builtin (1.47s)
```

I might have to do some digging later, after the other test failures are resolved, I guess.

**Comment by akerouanton** [comment]
@neersighted I just rebased onto master, hopefully the CI will be green this time.

**Review by neersighted — APPROVED** [review]
This LGTM to me overall. There are sustained concerns about proliferating multiple exported IP version enums in libnetwork (including your other open PR) -- however, that one can be changed with a different design, and this approach seems pragmatic to me. I'd like another maintainer to review before merge, however.

**Review by corhere — CHANGES_REQUESTED** [review]
I am in the process of cleaning up the IPAM packages (#44827 + [WIP](https://github.com/moby/moby/compare/master...corhere:moby:libnet/local-scope-only?expand=1)) to get rid of global mutable state and dead code. (Fun fact: `ipam.NewAllocator` is only ever called with `nil` datastores! All the datastore-related code can be thrown out.) Let's put this PR on hold until it can be rebased on top of the IPAM cleanup & refactor work, as that will open up a lot more flexibility to implement on-demand subnet splitting more cleanly.

I have spotted a much less invasive short-term mitigation for the IPv6 memory exhaustion issue which would be feasible to backport to 23.0: set a cap on the number of networks `ipamutils.splitNetworks` will return. 2^24 networks per address space should be well beyond what could be allocated without running into other limitations, and small enough to not exhaust all memory.

File: libnetwork/ipam/allocator.go
Comment (akerouanton) [inline libnetwork/ipam/allocator.go:49]: I'm not super happy with that: an address space with mixed address families feels weird. On the other hand `ReleasePool` currently relies on the pool's string identifier to determine what address space is involved.

@corhere @thaJeztah I can make a follow-up PR to make `addrSpace` handle a single address family if you think it'd be better.

File: libnetwork/ipam/structures.go
Comment (corhere) [inline libnetwork/ipam/structures.go:45]: I still don't like this whole concept of holding the mutable iterator state with the collection being iterated over, and having this "can be reset" flag to paper over the issues stemming from that design. For one, there is no clean way to iterate over a permutation of the collection without repeating values. A concrete toy example might better illustrate what I'm trying to say.

Consider the following iterator, where `^` signifies the current position.
```
[ 1 2 3 4 5 ]
    ^
```
With an API similar to `Subnetter`, attempting to iterate over the whole collection starting from the current position would yield the sequence `2 3 4 5 1 2 3 4 5`. If one could assume the values in the collection are all unique, the iterating code could remember that `2` was the first value iterated over and stop once it iterates over `2` again. But that would leave the current position incremented one past where iteration "stopped," a fencepost error.

I think a better design would be to separate the iterator from the collection. Make the collection immutable and the iterator ephemeral, yielding exactly one permutation of the collection. Here's a sketch of what that could look like:
```go
// Cursor returns a cursor positioned at the start of the pool.
func (SubnetPool) Cursor() SubnetCursor

// Next returns a cursor moved forward by one position.
func (SubnetCursor) Advance() SubnetCursor
// Iter returns a new iterator over a permutation of the pool starting from the cursor position.
func (SubnetCursor) Iter() *SubnetIter

// Next advances the iterator cursor by one and returns the subnet at that position.
// If advancing the cursor would position it at a subnet
// which has already been returned by a prior call to Next,
// the cursor is not advanced and ok=false is returned.
func (*SubnetIter) Next() (v netip.Prefix, ok bool)
// Cursor returns a cursor for the current position.
func (SubnetIter) Cursor() SubnetCursor
```

The idea is to use the cursor in the same way as `predefinedStartIndex`: remember the last place iteration terminated so the next round of iteration knows where to start and stop.

File: libnetwork/ipamutils/utils.go
Comment (corhere) [reply libnetwork/ipamutils/utils.go:34]: I think the code can be structured such that no IPVersion constants are needed.

**Review by corhere — CHANGES_REQUESTED** [review]
What's with all the "follow-up PR" talk? Is there some time pressure to get large pools fixed ASAP?

**Comment by akerouanton** [comment]
So, after some more tests, it turns out this PR makes clear that either the way we check for overlapping subnets is quite naive (it works well for IPv4, but not for IPv6) and/or the current `Subnetter` is the wrong abstraction.

For instance, if a user specify a default address pool with the base CIDR `10.0.0.0/8` and a subnet size of 30 in their daemon.json, and set `bip` config param as `10.0.0.1/8`, when the user want to create a new custom network the daemon has to iterate over all 2^(30-8) subnets (~ 4 millions) before returning an error. This takes ~ 0.5s (on my computer) and the issue is not really noticeable. Unfortunately, with IPv6 the worst case is so much bigger than it'd become unusable: with a /64 address pool split into /120 subnets, the current implementation has to iterate over 2^56 possible subnets (~ 7 * 10^16) before returning an error.

<details>

```console
$ cat /etc/docker/daemon.json
{
	"log-level": "info",
	"bip": "10.0.0.1/8",
	"fixed-cidr": "10.0.0.0/8",
	"default-address-pools": [
		{ "base": "10.0.0.0/8", "size": 30 }
	]
}
$ docker network create testnet1
```
</details>

Now, another case where that same bug happens. Two default address pools are defined: the first one yielding a single subnet and matching the bip/fixed-cidr params, and another, random, address pool (eg. `10.0.0.0/8` with size 24). Now if the user creates a network with a subnet matching the 2nd address pool CIDR, and then try to create a 3rd network with no subnet specified, all subnets from the 2nd address pool are enumerated before `(*addrSpace).allocatePredefinedPool()` returns an error.

<details>

```console
$ cat /etc/docker/daemon.json
{
	"bip": "172.19.0.1/24",
	"fixed-cidr": "172.19.0.0/24",
	"default-address-pools": [
		{ "base": "172.19.0.0/24", "size": 24 },
		{ "base": "10.0.0.0/8", "size": 24 }
	]
}
$ docker network create --subnet=10.0.0.0/8 testnet1
$ docker network create testnet2
```
</details>

Since `addrSpace` is locked during the "sometimes almost endless" loop (in `(*addrSpace).allocatePredefinedPool()`), the network controller can neither request nor release addresses and subnets in the mean time.

I can only come up with these two options to fix this issue:

* Disallow bip/fixed-cidr to overlap with one of the default address pools, and disallow custom networks with a user defined subnet to collide with one of the default address pools. We might break things if we do so, so probably not a good idea.
* Make the Subnetter aware of what subnets are currently in use.
  * Unfortunately, just moving its internal offset to ignore custom subnets won't be enough since the whole subnetter might be reset (and it's currently always reset at start up because of the way the default bridge is restored). Once reset, we're back to square one.
  * The other option is to merge `Subnetter` into `addrSpace`, to make it able to yield new subnets based on default address pools configured and user-supplied custom subnets. This poses the question of what data structure to use for that purpose. I can't think of anything else than LC/LPC/radix tries, but I'm wondering if a simpler data structure could help solve this issue.

**Comment by polarathene** [comment]
<details>
<summary>Earlier response</summary>

> For instance, if a user specify a default address pool with the base CIDR `10.0.0.0/8` and a subnet size of 30 in their `daemon.json`, and set `bip` config param as `10.0.0.1/8`, when the user want to create a new custom network the daemon has to iterate over all `2^(30-8)` subnets (~ 4 millions) before returning an error.

Realistically that many subnets is not going to be used. Current defaults AFAIK has a much smaller amount of subnets available in the pool? (_IIRC, this is poorly documented though_)

The issue here though is:
- Address pool with `/8` base to be subdivided into `/30` subnets.
- Any other network setting an explicit subnet that overlaps (the `/8`).

Only one of those configurations is valid to support at a time, and could be detected when starting the daemon?
- If the `/8` is already reserved for the `docker0` bridge, then the address pool base of `/8` or smaller would be invalid before considering `size`.
- If address pools come first, you've already reserved the entire `/8` subnet, again size can be ignored since it's not relevant, assigning a `/8` for `docker0` should fail.

Same logic applies for IPv6.

---

> Now if the user creates a network with a subnet matching the 2nd address pool CIDR, and then try to create a 3rd network with no subnet specified, all subnets from the 2nd address pool are enumerated before `(*addrSpace).allocatePredefinedPool()` returns an error.

This is a bit more interesting. But same concern, it's not that pragmatic to iterate through unallocated pools like that when realistically the pools already convey reservation of the `base` that is split into subnets of `size`.

This check is only important when creating a new network correct? Networks that are created persist their config with the assigned subnet, and pragmatically, how large does the number of assigned subnets get? It isn't likely to be that expensive to track with a hashmap is it?

**EDIT:** I might have misunderstood, is this about:
- `docker network create --subnet 10.0.0.0/8` (_not `10.0.0.0/24`?_)
- `docker network create` (_implicitly tries to assign a `10.0.0.0/24`?_)

Doesn't really change my feedback though either way.

</details>

---

> Disallow bip/fixed-cidr to overlap with one of the default address pools, and disallow custom networks with a user defined subnet to collide with one of the default address pools.
> We might break things if we do so, so probably not a good idea.

Why is it not viable to track what subnets are actually in use? The pool is just defining blocks of address space that can be assigned a subnet of `size` when one is not explicitly requested.

> The other option is to merge `Subnetter` into `addrSpace`, to make it able to yield new subnets based on default address pools configured and user-supplied custom subnets.
> This poses the question of what data structure to use for that purpose.
> I can't think of anything else than LC/LPC/radix tries, but I'm wondering if a simpler data structure could help solve this issue.

Seems we're on the same page :grinning: :+1: 

[Kubernetes uses bitmaps](https://github.com/kubernetes/kubernetes/issues/44918) apparently. Roaring compressed bitmaps were adopted as an improvement. There is some recent [related work on k8s](https://github.com/kubernetes/kubernetes/pull/116519), so I think they're still preferring that.

There's also a [nice article on investigating a few data structures for working with IPs](https://blog.apnic.net/2021/06/04/storing-and-retrieving-ip-prefixes-efficiently/) (Apnic), although I'm not sure how well any handle range queries to detect overlapping subnets :sweat_smile: 

---

I put something together with some rust code, data structure isn't simpler but I deferred that to existing libraries.
- Crate to [parse the string addresses into CIDR blocks](https://docs.rs/cidr/latest/cidr/index.html)
- Crate [implementing an R-Tree](https://github.com/tidwall/rtree.rs)

```rust
impl SubnetTree for SubnetManager<IPv6> {
    fn add(self: &mut Self, cidr_block: &str) {
      let subnet = Self::subnet_from_range(cidr_block);

      if let Some(overlapped) = self.tree.search(subnet).next() {
        println!("{} overlaps IP range with subnet {}", cidr_block, overlapped.data);
      } else {
        self.tree.insert(subnet, cidr_block.to_string());
        println!("Added subnet {}", cidr_block);
      }
    }
  }
```

Main logic is mostly that, and can be used like this:

```rust
fn main() {
  let mut subnets = SubnetManager::<IPv6>::new();
  let existing = ["fd00:feed:face:cafe::/64", "fd00:feed:face:beef::/64"];

  // These add without issue:
  for cidr_block in existing {
    subnets.add(cidr_block);
  }

  // These will fail due to overlap:
  subnets.add("fd00:feed:face:beef::/48");
  subnets.add("fd00:feed:face::/48");
}
```

Works with `IPv4` as well, although I just used two different trees for that. Each tree sets dimensions used by the number of octets (_4 for IPv4, 16 for IPv6_), which AFAIK simplifies the range queries (_I've not done any measurements_).

**Comment by akerouanton** [comment]
I'm going to close this PR in favor of https://github.com/moby/moby/pull/47768. As commented [here](https://github.com/moby/moby/pull/43033#issuecomment-1508703448), this implementation is a dead end.

**Pull request closed without merging**
